blockchain/chainio.go

```diff
@@ -1164,24 +1164,24 @@ func (b *BlockChain) initChainState() error {
 			// and add it to the block index.
 			node := &blockNodes[height]
 			initBlockNode(node, header, height)
-			if parent := b.bestNode; parent != nil {
-				node.parent = parent
-				node.workSum = node.workSum.Add(parent.workSum,
+			if tip != nil {
+				node.parent = tip
+				node.workSum = node.workSum.Add(tip.workSum,
 					node.workSum)
 			}
-			node.inMainChain = true
 			b.index.AddNode(node)
 
 			// This node is now the end of the best chain.
-			b.bestNode = node
+			b.bestChain.SetTip(node)
+			tip = node
 		}
 
-		// Ensure the resulting best node matches the stored best state
+		// Ensure the resulting best chain matches the stored best state
 		// hash.
-		if b.bestNode.hash != state.hash {
+		if b.bestChain.Tip().hash != state.hash {
```

Comment:
Good suggestion.  I was specifically trying to ensure the `bestChain` was set to tip here, but in conjunction with your other comment, I think it does indeed make more sense to check the tip local here and then `b.bestChain.SetTip(tip)` after it.